File: backend/src/main/java/com/vibecoding/rdq/service/ExternalIntegrationService.java
package com.vibecoding.rdq.service;

import com.vibecoding.rdq.dto.ExternalIntegrationResponse;
import com.vibecoding.rdq.entity.RDQ;
import org.springframework.stereotype.Service;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

/**
 * Service de gestion des intégrations externes
 * Génère les URLs pour les intégrations avec les applications externes
 */
@Service
public class ExternalIntegrationService {

    /**
     * Génère toutes les intégrations externes disponibles pour un RDQ
     */
    public List<ExternalIntegrationResponse> generateExternalIntegrations(RDQ rdq) {
        List<ExternalIntegrationResponse> integrations = new ArrayList<>();
        
        // Intégration email
        integrations.add(generateEmailIntegration(rdq));
        
        // Intégration Google Maps
        integrations.add(generateMapsIntegration(rdq));
        
        // Intégration Google Calendar
        integrations.add(generateCalendarIntegration(rdq));
        
        return integrations;
    }

    /**
     * Génère l'intégration email
     */
    private ExternalIntegrationResponse generateEmailIntegration(RDQ rdq) {
        String managerEmail = rdq.getManager() != null ? rdq.getManager().getEmail() : null;
        
        if (managerEmail == null || managerEmail.trim().isEmpty()) {
            return new ExternalIntegrationResponse("email", null, false, "Aucun email disponible");
        }

        try {
            String subject = URLEncoder.encode(
                String.format("RDQ #%d - %s", rdq.getIdRdq(), rdq.getTitre()),
                StandardCharsets.UTF_8
            );
            
            String body = URLEncoder.encode(buildEmailBody(rdq), StandardCharsets.UTF_8);
            String url = String.format("mailto:%s?subject=%s&body=%s", managerEmail, subject, body);
            
            return new ExternalIntegrationResponse(
                "email", 
                url, 
                true, 
                String.format("Envoyer un email à %s", managerEmail)
            );
        } catch (Exception e) {
            return new ExternalIntegrationResponse("email", null, false, "Erreur lors de la génération de l'email");
        }
    }

    /**
     * Génère l'intégration Google Maps
     */
    private ExternalIntegrationResponse generateMapsIntegration(RDQ rdq) {
        if (rdq.getAdresse() == null || rdq.getAdresse().trim().isEmpty() || 
            "DISTANCIEL".equals(rdq.getMode())) {
            String tooltip = "DISTANCIEL".equals(rdq.getMode()) ? 
                "Rendez-vous distanciel - pas d'adresse" : 
                "Aucune adresse disponible";
            return new ExternalIntegrationResponse("maps", null, false, tooltip);
        }

        try {
            String address = URLEncoder.encode(rdq.getAdresse(), StandardCharsets.UTF_8);
            String url = String.format("https://www.google.com/maps/search/?api=1&query=%s", address);
            
            return new ExternalIntegrationResponse(
                "maps", 
                url, 
                true, 
                "Ouvrir l'adresse dans Google Maps"
            );
        } catch (Exception e) {
            return new ExternalIntegrationResponse("maps", null, false, "Erreur lors de la génération du lien Maps");
        }
    }

    /**
     * Génère l'intégration Google Calendar
     */
    private ExternalIntegrationResponse generateCalendarIntegration(RDQ rdq) {
        if (rdq.getDateHeure() == null) {
            return new ExternalIntegrationResponse("calendar", null, false, "Aucune date/heure disponible");
        }

        try {
            LocalDateTime startDate = rdq.getDateHeure();
            LocalDateTime endDate = startDate.plusHours(2); // +2h par défaut

            String title = URLEncoder.encode(
                String.format("RDQ - %s", rdq.getTitre()),
                StandardCharsets.UTF_8
            );
            
            String details = URLEncoder.encode(buildCalendarDetails(rdq), StandardCharsets.UTF_8);
            
            String location = URLEncoder.encode(
                rdq.getAdresse() != null && !"DISTANCIEL".equals(rdq.getMode()) 
                    ? rdq.getAdresse() 
                    : "Rendez-vous distanciel",
                StandardCharsets.UTF_8
            );

            String dates = formatDateForGoogle(startDate) + "/" + formatDateForGoogle(endDate);
            
            String url = String.format(
                "https://calendar.google.com/calendar/render?action=TEMPLATE&text=%s&dates=%s&details=%s&location=%s",
                title, dates, details, location
            );
            
            return new ExternalIntegrationResponse(
                "calendar", 
                url, 
                true, 
                "Ajouter au calendrier Google"
            );
        } catch (Exception e) {
            return new ExternalIntegrationResponse("calendar", null, false, "Erreur lors de la génération de l'événement calendrier");
        }
    }

    /**
     * Construit le corps de l'email
     */
    private String buildEmailBody(RDQ rdq) {
        StringBuilder body = new StringBuilder();
        body.append("Bonjour,\n\n");
        body.append("Je vous contacte concernant le RDQ suivant :\n\n");
        body.append(String.format("• Titre : %s\n", rdq.getTitre()));
        
        if (rdq.getDateHeure() != null) {
            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("EEEE d MMMM yyyy 'à' HH:mm");
            body.append(String.format("• Date : %s\n", rdq.getDateHeure().format(formatter)));
        }
        
        body.append(String.format("• Mode : %s\n", rdq.getMode()));
        
        if (rdq.getAdresse() != null && !rdq.getAdresse().trim().isEmpty()) {
            body.append(String.format("• Adresse : %s\n", rdq.getAdresse()));
        }
        
        body.append("\nCordialement.");
        
        return body.toString();
    }

    /**
     * Construit les détails de l'événement calendrier
     */
    private String buildCalendarDetails(RDQ rdq) {
        StringBuilder details = new StringBuilder();
        details.append(String.format("RDQ #%d\n\n", rdq.getIdRdq()));
        
        if (rdq.getManager() != null) {
            details.append(String.format("Manager : %s\n", rdq.getManager().getNom()));
        }
        
        details.append(String.format("Mode : %s\n", rdq.getMode()));
        
        if (rdq.getDescription() != null && !rdq.getDescription().trim().isEmpty()) {
            details.append(String.format("\nDescription :\n%s", rdq.getDescription()));
        }
        
        if (rdq.getDescription() != null && !rdq.getDescription().trim().isEmpty()) {
            details.append(String.format("\n\nDescription complète :\n%s", rdq.getDescription()));
        }
        
        return details.toString();
    }

    /**
     * Formate une date pour Google Calendar (format ISO sans séparateurs)
     */
    private String formatDateForGoogle(LocalDateTime date) {
        return date.format(DateTimeFormatter.ofPattern("yyyyMMdd'T'HHmmss"));
    }
}

File: frontend-react/frontend/src/services/ExternalIntegrationService.ts
/**
 * Service de gestion des intégrations externes
 * Génère les URLs pour l'intégration avec les applications externes (email, maps, calendrier)
 */

import { RDQ } from '../types';

export interface ExternalActionData {
  type: 'email' | 'maps' | 'calendar';
  enabled: boolean;
  url?: string;
  tooltip: string;
}

export class ExternalIntegrationService {
  
  /**
   * Génère l'URL pour l'intégration email
   */
  static generateEmailUrl(rdq: RDQ): string | null {
    // Prioriser l'email du manager
    const email = rdq.manager?.email;
    
    if (!email) {
      return null;
    }

    const subject = encodeURIComponent(`RDQ #${rdq.idRdq} - ${rdq.titre}`);
    const body = encodeURIComponent(
      `Bonjour,\n\n` +
      `Je vous contacte concernant le RDQ suivant :\n\n` +
      `• Titre : ${rdq.titre}\n` +
      `• Date : ${new Date(rdq.dateHeure).toLocaleDateString('fr-FR', {
        weekday: 'long',
        year: 'numeric',
        month: 'long',
        day: 'numeric',
        hour: '2-digit',
        minute: '2-digit'
      })}\n` +
      `• Mode : ${rdq.mode}\n` +
      (rdq.adresse ? `• Adresse : ${rdq.adresse}\n` : '') +
      `\n` +
      `Cordialement.`
    );

    return `mailto:${email}?subject=${subject}&body=${body}`;
  }

  /**
   * Génère l'URL pour l'intégration Google Maps
   */
  static generateMapsUrl(rdq: RDQ): string | null {
    if (!rdq.adresse || rdq.mode === 'DISTANCIEL') {
      return null;
    }

    const address = encodeURIComponent(rdq.adresse);
    return `https://www.google.com/maps/search/?api=1&query=${address}`;
  }

  /**
   * Génère l'URL pour l'intégration Google Calendar
   */
  static generateCalendarUrl(rdq: RDQ): string | null {
    const startDate = new Date(rdq.dateHeure);
    const endDate = new Date(startDate.getTime() + (2 * 60 * 60 * 1000)); // +2h par défaut

    // Format ISO sans les millisecondes ni le Z final pour Google Calendar
    const formatDateForGoogle = (date: Date): string => {
      return date.toISOString().replace(/\.\d{3}Z$/, 'Z').replace(/[:-]/g, '').replace('Z', '');
    };

    const title = encodeURIComponent(`RDQ - ${rdq.titre}`);
    const details = encodeURIComponent(
      `RDQ #${rdq.idRdq}\n\n` +
      `Manager : ${rdq.manager?.nom || 'Non défini'}\n` +
      `Mode : ${rdq.mode}\n` +
      (rdq.description ? `\nDescription :\n${rdq.description}` : '') +
      (rdq.indications ? `\n\nIndications :\n${rdq.indications}` : '')
    );

    const location = rdq.adresse && rdq.mode !== 'DISTANCIEL' 
      ? encodeURIComponent(rdq.adresse)
      : encodeURIComponent('Rendez-vous distanciel');

    const dates = `${formatDateForGoogle(startDate)}/${formatDateForGoogle(endDate)}`;

    return `https://calendar.google.com/calendar/render?action=TEMPLATE&text=${title}&dates=${dates}&details=${details}&location=${location}`;
  }

  /**
   * Génère les données d'action pour un RDQ donné
   */
  static generateExternalActions(rdq: RDQ): ExternalActionData[] {
    const actions: ExternalActionData[] = [];

    // Action Email
    const emailUrl = this.generateEmailUrl(rdq);
    actions.push({
      type: 'email',
      enabled: !!emailUrl,
      url: emailUrl || undefined,
      tooltip: emailUrl 
        ? `Envoyer un email à ${rdq.manager?.email}`
        : 'Aucun email disponible'
    });

    // Action Maps
    const mapsUrl = this.generateMapsUrl(rdq);
    actions.push({
      type: 'maps',
      enabled: !!mapsUrl,
      url: mapsUrl || undefined,
      tooltip: mapsUrl 
        ? `Ouvrir l'adresse dans Google Maps`
        : rdq.mode === 'DISTANCIEL' 
          ? 'Rendez-vous distanciel - pas d\'adresse'
          : 'Aucune adresse disponible'
    });

    // Action Calendar
    const calendarUrl = this.generateCalendarUrl(rdq);
    actions.push({
      type: 'calendar',
      enabled: !!calendarUrl,
      url: calendarUrl || undefined,
      tooltip: calendarUrl 
        ? 'Ajouter au calendrier Google'
        : 'Impossible de générer l\'événement calendrier'
    });

    return actions;
  }

  /**
   * Ouvre une URL externe dans une nouvelle fenêtre/onglet
   */
  static openExternalUrl(url: string): void {
    window.open(url, '_blank', 'noopener,noreferrer');
  }

  /**
   * Gestionnaire d'événement pour les actions externes
   */
  static handleExternalAction(action: ExternalActionData, rdq: RDQ): void {
    if (!action.enabled || !action.url) {
      console.warn(`Action ${action.type} non disponible pour RDQ #${rdq.idRdq}`);
      return;
    }

    // Log pour analytics (optionnel)
    console.log(`External action triggered: ${action.type} for RDQ #${rdq.idRdq}`);

    // Ouvrir l'URL
    this.openExternalUrl(action.url);
  }

  /**
   * Validation des données RDQ pour les intégrations externes
   */
  static validateRdqForExternalActions(rdq: RDQ): {
    hasEmail: boolean;
    hasAddress: boolean;
    hasDateTime: boolean;
    warnings: string[];
  } {
    const warnings: string[] = [];
    
    const hasEmail = !!(rdq.manager?.email);
    const hasAddress = !!(rdq.adresse && rdq.mode !== 'DISTANCIEL');
    const hasDateTime = !!(rdq.dateHeure);

    if (!hasEmail) {
      warnings.push('Aucun email de manager disponible pour l\'intégration email');
    }

    if (!hasAddress && rdq.mode === 'PRESENTIEL') {
      warnings.push('Adresse manquante pour un RDQ en présentiel');
    }

    if (!hasDateTime) {
      warnings.push('Date/heure manquante pour l\'intégration calendrier');
    }

    return {
      hasEmail,
      hasAddress,
      hasDateTime,
      warnings
    };
  }
}

export default ExternalIntegrationService;